jupyterlab_gcedetails/src/components/resource_utilization_charts.tsx

```diff
@@ -0,0 +1,102 @@
+import * as React from 'react';
+import { ServerWrapper } from './server_wrapper';
+import { AreaChartWrapper } from './chart_wrapper';
+import { Utilization, STYLES } from '../data';
+
+const AREA_CHART_BLUE = {
+  stroke: '#15B2D3',
+  fill: '#15B2D3',
+};
+
+const AREA_CHART_ORANGE = {
+  stroke: '#ff7f01',
+  fill: '#ff7f01',
+};
+
+const UTILIZATION_CHART_PROPERTIES = {
+  areaChartProps: {
+    height: 75,
+    width: 350,
+  },
+  yAxisProps: {
+    domain: [0, 100],
+  },
+  cartesianGridProps: {
+    horizontalPoints: [25, 50, 75],
+    vertical: false,
+  },
+};
+
+interface Props {
+  serverWrapper: ServerWrapper;
+}
+
+interface State {
+  data?: Utilization[];
+  receivedError: boolean;
+}
+
+export class ResourceUtilizationCharts extends React.Component<Props, State> {
+  private readonly refreshInterval: number;
+  private readonly NUM_DATA_POINTS = 20;
+  private readonly REFRESH_INTERVAL = 1000;
+  constructor(props: Props) {
+    super(props);
+    const data = [];
+    for (let i = 0; i < this.NUM_DATA_POINTS; i++) {
+      data.push({ cpu: 0, memory: 0 });
+    }
+    this.state = {
+      data,
+      receivedError: false,
+    };
+    this.refreshInterval = window.setInterval(() => {
```

Comment:
I would suggest creating this interval when the component mounts rather than the constructor. In general that seems like the suggested place to initiate any calls to a backend.

https://reactjs.org/docs/react-component.html#componentdidmount

Reply:
I think since `refreshInterval`  is readonly type, it can't be assigned in `componentDidMount` which is why I think it's assigned in the constructor in the original implementation in `details_widget.tsx` as well.

Reply:
That is correct. However in `details_widget.tsx`, the logic created the interval but used the `shouldRefresh` state variable to determine whether the call to the server should be made, which would initially be false. In this case, the visibility of the chart determines that the data should be refreshed. You could drop the `readonly` modifier from the property and assign a new interval in the `componentDidMount`, and clear it in `componentWillUnmount`.